From caafe296bdf4093d224adf3413dd17f59d2b45dc Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ond=C5=99ej=20Caletka?= <ondrej@caletka.cz>
Date: Tue, 28 Jan 2014 12:30:51 +0100
Subject: [PATCH 2/5] Support for future SSHFP digest algorithms.

When a SSHFP record is found in DNS, which has digest type other than
SHA1 or SHA256, whole SSHFP matching is aborted. This patch fixes it.
---
 dns.c | 21 +++++++--------------
 1 file changed, 7 insertions(+), 14 deletions(-)

diff --git a/dns.c b/dns.c
index 0c2e9b2..817dfa9 100644
--- a/dns.c
+++ b/dns.c
@@ -200,7 +200,7 @@ verify_host_key_dns(const char *hostname, struct sockaddr *address,
 
 	u_int8_t hostkey_algorithm;
 	u_int8_t hostkey_digest_type = SSHFP_HASH_RESERVED;
-	u_char *hostkey_digest;
+	u_char *hostkey_digest = NULL;
 	u_int hostkey_digest_len;
 
 	u_int8_t dnskey_algorithm;
@@ -240,14 +240,6 @@ verify_host_key_dns(const char *hostname, struct sockaddr *address,
 		    fingerprints->rri_nrdatas);
 	}
 
-	/* Initialize default host key parameters */
-	if (!dns_read_key(&hostkey_algorithm, &hostkey_digest_type,
-	    &hostkey_digest, &hostkey_digest_len, hostkey)) {
-		error("Error calculating host key fingerprint.");
-		freerrset(fingerprints);
-		return -1;
-	}
-
 	if (fingerprints->rri_nrdatas)
 		*flags |= DNS_VERIFY_FOUND;
 
@@ -266,15 +258,15 @@ verify_host_key_dns(const char *hostname, struct sockaddr *address,
 
 		if (hostkey_digest_type != dnskey_digest_type) {
 			hostkey_digest_type = dnskey_digest_type;
-			free(hostkey_digest);
+			if (hostkey_digest)
+				free(hostkey_digest);
 
 			/* Initialize host key parameters */
 			if (!dns_read_key(&hostkey_algorithm,
 			    &hostkey_digest_type, &hostkey_digest,
 			    &hostkey_digest_len, hostkey)) {
-				error("Error calculating key fingerprint.");
-				freerrset(fingerprints);
-				return -1;
+				debug("Error calculating key fingerprint.");
+				continue;
 			}
 		}
 
@@ -292,7 +284,8 @@ verify_host_key_dns(const char *hostname, struct sockaddr *address,
 		free(dnskey_digest);
 	}
 
-	free(hostkey_digest); /* from key_fingerprint_raw() */
+	if (hostkey_digest)
+		free(hostkey_digest); /* from key_fingerprint_raw() */
 	freerrset(fingerprints);
 
 	if (*flags & DNS_VERIFY_FOUND) {
-- 
1.8.3.2

